### ISSUE TITLE
Standalone Server + Client example

### ISSUE THREAD

**Issue by kdvolder** [eclipse/sprotty#96]
We are interested in using sprotty in the context of LSP. We have already got (several) language servers that are built using lsp4j, lsp4e and vscode language server client, and theia.

I found this:
https://github.com/eclipse/sprotty/tree/master/examples

I think those are just showing some 'toy' 'client-only' setups (being toy is great actually, but these examples do not help me understand how a client<->server setup is done.

I also found this:

https://github.com/TypeFox/theia-xtext-sprotty-example

Looks like that does have a server component. And so I will try to work through that. But it isn't quite what we are looking for. (We do not really need/want to be building a DSL with xtext, since we already have a language server). In the outset we also don't necessarily want to integrate with Theia (eventually I am sure we do, but our first priority is just to understand the client<->server setup, more or less independently from a specific IDE context.)

So... what would be nice is a simple/toy example of a setup that has a standalone server and client, that can be run and then opened in a web browser. (I.e. something similar to the simple example () but with a server component connected to the client).

What would be nice is an example that just demonstrates how to get started with creating/running a (standalone) model-server and connecting to a client to render the model(s) it serves.

Does something like that exist? 

I am guessing the answer is no. In that case then... could you provide some pointers on how I could go about creating it for myself?). E.g. let's say I'd start with one of the simple examples, like the class diagramming. How would I go about moving the model creation out of the client code and get it from a server instead. And how would I go about setting up the corresponding server? 

I would really appreciate any help, and... if I succeed, I'd be willing to share the resulting 'sample code' in a public repo, or even contribute it back to sprotty via a PR, as I am sure it would be helpful to others too.

**Comment by kdvolder** [comment]
A bit disappointed in lack of response on this ticket.

Anyhow... it took a lot of guessing and trial and error but we got a version of the 'circles' examples working with a standalone server. Might be of interest to someone. The server in our case is implemented in Java using spring-boot websocket support.

It can be found here: https://github.com/kdvolder/sprotty-server-example

Still in a bit of mess (just got it working yesterday out a lot of copy-pasted bits) and there is no explaniations or readmes. But perhaps it is useful to someone looking for similar answer than us anyways.

Sprotty is indeed impressive, and the api design for implementing server is pretty good once you manage to figure out how its intended to be used... But the lack of documentation and examples makes it rather inaccessible to use by an 'outsider'. So I do hope you'll consider doing something about that. (Yes, I know... its a lot of work to make good docs / examples :-).

PS: Couldn't figure out how this really ties in with the LSP protocol so our example just uses raw webockets as transport layer.

**Comment by spoenemann** [comment]
Sorry, I've been busy last week.

> What would be nice is an example that just demonstrates how to get started with creating/running a (standalone) model-server and connecting to a client to render the model(s) it serves.
> Does something like that exist?
> I am guessing the answer is no.

Correct. We work a lot with Theia and LSP (esp. Xtext), and so all our Sprotty client-server examples and projects so far are based on that.

> it took a lot of guessing and trial and error but we got a version of the 'circles' examples working with a standalone server.

Cool, looks good! There is also a `DiagramServerEndpoint` class for Websocket connection, do you think it's useful?

> But the lack of documentation and examples makes it rather inaccessible to use by an 'outsider'. So I do hope you'll consider doing something about that.

Yes, that's been on my agenda for some time already. But somehow a lot of other (seemingly more important) things keep kicking in...

> Couldn't figure out how this really ties in with the LSP protocol so our example just uses raw webockets as transport layer.

Our approach is to extend the JSON-RPC protocol with an additional method `diagram/accept` that takes an `ActionMessage` as parameter. This new message is valid both from client to server and from server to client. You have to teach both sides how to forward those messages to the respective Sprotty parts. For Theia we have the [sprotty-theia](https://github.com/eclipse/sprotty-theia) package that does the client-side message handling (and manages Sprotty widgets in Theia), and for Xtext LSP we have the [org.eclipse.sprotty.xtext](https://github.com/eclipse/sprotty-server/tree/master/org.eclipse.sprotty.xtext) module. The [theia-xtext-sprotty-example](https://github.com/TypeFox/theia-xtext-sprotty-example) you already mentioned demonstrates this.

In summary: how to use Sprotty with LSP highly depends on which client IDE and which language server you are using. In some cases it might be very difficult to add additional protocol methods, so an alternative solution may be necessary.

**Comment by kdvolder** [comment]
> There is also a DiagramServerEndpoint class for Websocket connection, do you think it's useful?

It was useful as a 'example' but not to use "as is". (For us anyway, maybe it could be for someone else though). We didn't use it because it's dependency on javax.WebSocket. Our language servers are implemented in spring so its more natural for us to use spring's support for implementing websocket endpoints. Spring provides an abstraction that makes that easy and decouples from the underlying server implementation. Not sure how javax.Websocket fits into that exactly. (It may be there somewhere under the hood, depending on which concrete server / websocket implementation you choose to use, but as a user of the framework you woulnd't touch/implement javax.WebSocket directly).

Anyhow... so while we couldn't / didn't use it directly, it was still useful to look at the code and to understand what we needed to do with our own websocket endpoint implementation.

> In summary: how to use Sprotty with LSP highly depends on which client IDE and which language server you are using

I see, that kind of defeats the purpose of LSP a bit perhaps? Seems like just using websockets may be the easier route which actually requires less custom implementation bits that have to be repeated for every different client environment?

**Comment by spoenemann** [comment]
>> In summary: how to use Sprotty with LSP highly depends on which client IDE and which language server you are using

> I see, that kind of defeats the purpose of LSP a bit perhaps? Seems like just using websockets may be the easier route which actually requires less custom implementation bits that have to be repeated for every different client environment?

Yes, the way we have chosen is to extend the protocol, but that means it only works with that specific combination of IDE client and language server. Opening an additional communication channel would be more portable.

**Comment by spoenemann** [comment]
Now there is an official server + client example:
https://github.com/eclipse/sprotty-server/tree/master/examples

Run `./gradlew runExamples` to launch the server.
